docs/architecture/adr-059-evidence-composition-and-lifecycle.md

```diff
@@ -0,0 +1,256 @@
+# ADR 059: Evidence Composition and Lifecycle
+
+## Changelog
+
+- 04/09/2020: Initial Draft (Unabridged)
+- 07/09/2020: First Version
+
+## Scope
+
+This document is designed to collate together and surface some predicaments involving evidence in Tendermint: both it's composition and lifecycle. It then aims to find a solution to these. The scope does not extend to the verification nor detection of certain types of evidence but concerns itself mainly with the general form of evidence and its lifecycle.
+
+## Background
+
+For a long time `DuplicateVoteEvidence`, formed in the consensus engine, was the only evidence Tendermint had. It was produced whenever two votes from the same validator in the same round
+was observed and thus it was made in a per validator way. It was predicted that there may come more forms of evidence and thus `DuplicateVoteEvidence` was used as the model for the `Evidence` interface and also for the form of the evidence data sent to the application. It is important to note that Tendermint concerns itself just with the detection and reporting of evidence and it is the responsibility of the application to exercise punishment.
+
+```go
+type Evidence interface { //existing
+  Height() int64                                     // height of the offense
+  Time() time.Time                                   // time of the offense
+  Address() []byte                                   // address of the offending validator
+  Bytes() []byte                                     // bytes which comprise the evidence
+  Hash() []byte                                      // hash of the evidence
+  Verify(chainID string, pubKey crypto.PubKey) error // verify the evidence
+  Equal(Evidence) bool                               // check equality of evidence
+
+  ValidateBasic() error
+  String() string
+}
+```
+
+```go
+type DuplicateVoteEvidence struct {
+  VoteA *Vote
+  VoteB *Vote
+
+  timestamp time.Time // taken from the block time
+}
+```
+
+Tendermint has now introduced a new type of evidence to protect light clients from being attacked. This `LightClientAttackEvidence` is vastly different to `DuplicateVoteEvidence` in that it is physically a much different size containing a complete signed header and validator set. It is formed within the light client, not the consensus engine and requires a lot more information from state to verify (`VerifyLightClientAttack(commonHeader, trustedHeader *SignedHeader, commonVals *ValidatorSet)`  vs `VerifyDuplicateVote(chainID string, pubKey PubKey)`). Finally it batches validators together as opposed to having individual evidence. This evidence stretches the existing mould that was used to accomodate new types of evidence and has thus caused us to reconsider how evidence should be formatted and processed.
+
+```go
+type LightClientAttackEvidence struct { // proposed struct in spec
+  ConflictingBlock *LightBlock
+  CommonHeight int64
```

Comment:
`CommonHeight` is the last height in common between the... witness and provider?

Reply:
The witness and the primary. Both of which are providers

Reply:
Ah, got it, thanks. Maybe: 

```suggestion
  CommonHeight int64 // the last height at which the primary provider and the witness provider agree
```